Add field device feature and specific field device objects in new SwzDeviceFeed

j-d-b opened the pull request:
This draft PR addresses #191 and #195 by defining models for smart work zone field devices and integrating them into the WZDx specification through a new feed called the [`SwzDeviceFeed`](https://github.com/usdot-jpo-ode/wzdx/blob/field-devices/spec-content/objects/SwzDeviceFeed.md).

Specifically, this PR adds a new GeoJSON feature type called [`FieldDeviceFeature`](https://github.com/usdot-jpo-ode/wzdx/blob/field-devices/spec-content/objects/FieldDeviceFeature.md). The `FieldDeviceFeature` can describe several types of field devices as enumerated by the new [`FieldDeviceType`](https://github.com/usdot-jpo-ode/wzdx/blob/field-devices/spec-content/enumerated-types/FieldDeviceType.md) enumerated type and more can be added as the specification evolves. 

Previously, WZDx only defined one type of feature, the [`RoadEventFeature`](https://github.com/usdot-jpo-ode/wzdx/blob/field-devices/spec-content/objects/RoadEventFeature.md), which occurs within the [`WZDxFeed`](https://github.com/usdot-jpo-ode/wzdx/blob/field-devices/spec-content/objects/WZDxFeed.md). To not modify the `WZDxFeed` which is designed for providing work zone road event information to the traveling public, the `FieldDeviceFeature` occurs within a new [`SwzDeviceFeed`](https://github.com/usdot-jpo-ode/wzdx/blob/field-devices/spec-content/objects/SwzDeviceFeed.md) (see #191 for background discussion) designed for providing deployed field device information.

The only effect to the `WZDxFeed` is that the `RoadEventFeedInfo` and `RoadEventDataSource` objects were renamed to `FeedInfo` and `FeedDataSource` as they now apply to more than just road events. Note this has no change in the structure or content of a WZDx feed.

## New Objects
Object | Description
--- | ---
[ArrowBoard](https://github.com/usdot-jpo-ode/wzdx/blob/field-devices/spec-content/objects/ArrowBoard.md) | An electronic, connected arrow board which can display an arrow pattern to direct traffic.
[Camera](https://github.com/usdot-jpo-ode/wzdx/blob/field-devices/spec-content/objects/Camera.md) | A camera device deployed in the field, capable of capturing still images.
[DynamicMessageSign](https://github.com/usdot-jpo-ode/wzdx/blob/field-devices/spec-content/objects/DynamicMessageSign.md) | An electronic traffic sign deployed on the roadway, used to provide information to travelers.
[FlashingBeacon](https://github.com/usdot-jpo-ode/wzdx/blob/field-devices/spec-content/objects/FlashingBeacon.md) | A flashing beacon light of any form (e.g. trailer-mounted, vehicle), used to indicate something or capture driver attention.
[FieldDeviceFeature](https://github.com/usdot-jpo-ode/wzdx/blob/field-devices/spec-content/objects/FieldDeviceFeature.md) | The GeoJSON `Feature` container object for a deployed field device.
[FieldDeviceCoreDetails](https://github.com/usdot-jpo-ode/wzdx/blob/field-devices/spec-content/objects/FieldDeviceCoreDetails.md) | The core details—both configuration and current state—of a field device that are shared by all types of field devices.
[HybridSign](https://github.com/usdot-jpo-ode/wzdx/blob/field-devices/spec-content/objects/HybridSign.md) | A hybrid sign that contains static text (e.g. on an aluminum sign) along with a single electronic message display, used to provide information to travelers.
[LocationMarker](https://github.com/usdot-jpo-ode/wzdx/blob/field-devices/spec-content/objects/LocationMarker.md) | Describes any GPS-enabled ITS device that is placed at a point on a roadway to dynamically know the location of something (often the beginning or end of a work zone).
[MarkedLocation](https://github.com/usdot-jpo-ode/wzdx/blob/field-devices/spec-contentobjects/MarkedLocation.md) | Describes a specific location where a [LocationMarker](https://github.com/usdot-jpo-ode/wzdx/blob/field-devices/spec-content/objects/LocationMarker.md) is placed, such as the start or end of a work zone road event.
[SwzDeviceFeed](https://github.com/usdot-jpo-ode/wzdx/blob/field-devices/spec-content/objects/SwzDeviceFeed.md) | The root (highest level) object of a smart work zone devices feed.
[TrafficSensor](https://github.com/usdot-jpo-ode/wzdx/blob/field-devices/spec-content/objects/TrafficSensor.md) | A traffic sensor deployed on a roadway which captures traffic metrics (e.g. speed, volume, occupancy) over a collection interval.
[TrafficSensorLaneData](https://github.com/usdot-jpo-ode/wzdx/blob/field-devices/spec-content/objects/TrafficSensorLaneData.md) | Data for a single lane within a [RoadEvent](https://github.com/usdot-jpo-ode/wzdx/blob/field-devices/spec-content/objects/RoadEvent.md) measured by a [TrafficSensor](https://github.com/usdot-jpo-ode/wzdx/blob/field-devices/spec-content/objects/TrafficSensor.md) deployed on the roadway.

## New enumerated types
Enumerated Type | Description
--- | ---
[ArrowBoardPattern](https://github.com/usdot-jpo-ode/wzdx/blob/field-devices/spec-content/enumerated-types/ArrowBoardPattern.md) | A list of options for the posted pattern on an [ArrowBoard](https://github.com/usdot-jpo-ode/wzdx/blob/field-devices/spec-content/objects/ArrowBoard.md).
[FlashingBeaconFunction](https://github.com/usdot-jpo-ode/wzdx/blob/field-devices/spec-content/enumerated-types/FlashingBeaconFunction.md) | Describes options for what a [FlashingBeacon](https://github.com/usdot-jpo-ode/wzdx/blob/field-devices/spec-content/objects/FlashingBeacon.md) is being used to indicate.
[HybridSignDynamicMessageFunction](https://github.com/usdot-jpo-ode/wzdx/blob/field-devices/spec-content/enumerated-types/HybridSignDynamicMessageFunction.md) | Describes options for the function of the variable message displayed by the electronic display on a [HybridSign](https://github.com/usdot-jpo-ode/wzdx/blob/field-devices/spec-content/objects/HybridSign.md).
[FieldDeviceStatus](https://github.com/usdot-jpo-ode/wzdx/blob/field-devices/spec-content/enumerated-types/FieldDeviceStatus.md) | The operational status of a field device.
[FieldDeviceType](https://github.com/usdot-jpo-ode/wzdx/blob/field-devices/spec-content/enumerated-types/EventStatus.md) | The type of a field device.
[MarkedLocationType](https://github.com/usdot-jpo-ode/wzdx/blob/field-devices/spec-content/enumerated-types/MarkedLocationType.md) | Describes options for what a [MarkedLocation](https://github.com/usdot-jpo-ode/wzdx/blob/field-devices/spec-content/objects/MarkedLocation.md) can mark, such as the start or end of a road event.

## WZDx Feeds
Feed Name | Description | Producer | Consumer | Uses | Content
--- | --- | --- | --- | --- | ---
`WZDxFeed` (existing) | Provides information about events ocurring on roadways (called "road events") such as work zones that impact the characteristics of the roadway and involve a change the roadway from the default state (e.g. a lane closure). The `WZDxFeed` is the original work zone data exchange feed, focused on describing high-level information about road work. | Agencies responsible for managing roadways and road work, typically state and local DOTs. | Traveling public via third parties such as mapping companies and CAVs. | Route planning; increased awareness; "put work zones on the map" | Work zone and detour road events (see [RoadEventFeature](/spec-content/objects/RoadEventFeature.md)).
`SwzDeviceFeed` (new) | Provides information (location, status, live data) about field devices deployed on the roadway in work zones. | Smart work zone equipment manufacturers or vendors. | Agencies responsible for managing roadways and permitting work, typically state and local DOTs who are often interested in producing a `WZDxFeed`. Third-parties such as mapping companies and CAVs may also be interested in field device information. | Simplifies design process for agencies wanting to interface with equipment manufacturers; aids in dynamically generating a WZDx feed with accurate information; reduces effort for manufacturers to conform to different agencies requirements. | Field devices (see [FieldDeviceFeature](/spec-content/objects/FieldDeviceFeature.md)).

## Data Flow

```
                                                                                 |
                                                                                 | Various
                                                                                 | Additional Inputs
                                                                                 v
                                                                       +------------------+            +----------------+
                                                                       |                  |    WZDx    |                |
                                                                       |     DOT or       |    Feed    |    Traveling   |
                                                           +---------> |  Similar Agency  | ---------> |     Public     |
  Deployed Field    Lower-Level   +------------------+     |           |                  |            |                |
     Devices       Communication  |                  |     |           +------------------+            +----------------+
+---+ +---+ +---+    Protocols    |    Equipment     | ----+  Swz                |
|   | |   | |   |  -------------> |   Manufacturer   |       Device              | WZDxFeed                     ^
+---+ +---+ +---+                 |    or Vendor     | ----+  Feed               v                              |
                                  |                  |     |           +------------------+                     |
                                  +------------------+     |           |   Third-Party    |                     |
                                                           +---------> |   Stakeholders   | --------------------+
                                                                       |  (e.g. mapping   |
                                                                       |  companies, CAVs)|
                                                                       +------------------+
```

This PR is a draft for demonstration of the new models and how they will be integrated into the WZDx specification. The PR is a draft as it can not be finalized or merged until other v4.0 PRs are merged to the release branch (`v4.0-release`) as it would create a huge amount of unresolvable merge conflicts and a lot of manual work. This PR does not update examples files or the object diagram. It does include fully complete documentation for all the new objects and a new JSON schema for the new smart work zone device feed.

j-d-b commented:
Following a comment from @sknick-iastate in the SWZ Subgroup meeting on 2021-09-27, it is clear that there should be an equivalent of the [RoadEventDataSource](https://github.com/usdot-jpo-ode/wzdx/blob/main/spec-content/objects/RoadEventDataSource.md) for field devices.

Since the RoadEventDataSource is becoming generic in #209, it can be used for field device features. Thus if #209 passes, the only change needed in this PR is for a `data_source_id` property needs to be added to the [FieldDeviceCoreDetails](https://github.com/usdot-jpo-ode/wzdx/blob/field-devices/spec-content/objects/FieldDeviceCoreDetails.md) object. This change is pending following resolution of #209.

j-d-b commented:
Following spec update co-chairs discussion on 2021-09-30 and swz co-chairs discussion on 2021-10-01 regarding #191, this PR was updated to include the FieldDeviceFeature in a new feed.

chuehlien reviewed (commented):
@j-d-b good work! I made some small formatting/typo changes and added some comments - please review when you get a chance. Thanks.

j-d-b commented on spec-content/objects/TrafficSensor.md:
```diff
@@ -0,0 +1,20 @@
+`collection_interval_start_date` | String; [date-time](https://tools.ietf.org/html/draft-handrews-json-schema-validation-01#section-7.3.1) | The UTC date and time where the TrafficSensor data began being collected at. The averages and totals contained in the TrafficSensor data apply to the inclusive interval of `collection_interval_start_date` to `collection_interval_end_date`. | Required |
+`collection_interval_end_date` | String; [date-time](https://tools.ietf.org/html/draft-handrews-json-schema-validation-01#section-7.3.1) | The UTC date and time where the TrafficSensor collection interval ended. The averages and totals contained in the TrafficSensor data apply to the inclusive interval of `collection_interval_start_date` to `collection_interval_end_date`. | Required |
+`average_speed_kph` | Positive Integer | The average speed of vehicles across all lanes over the collection interval in kilometers per hour. | Optional | 
+`volume_vph` | Integer | The rate of vehicles passing by the sensor during the collection interval in vehicles per hour. | Optional |
```
Thank you for finding this issue. I updated all traffic metrics (speed, volume, occupancy) to be type `Integer (>= 0)`. Occupancy type was "Number" in the markdown file but an integer in the schema and zero should be allowed for all, so not strictly positive integers.

j-d-b replied:
I was torn whether to write out "greater than or equal to" or use ">=". @chuehlien what do you think?

chuehlien replied:
Just found out that we could do `$\ge$` in markdown to get ≥! Maybe we can try that?

j-d-b replied:
Yes, great idea, let's do it.

j-d-b commented:
> I made some small formatting/typo changes and added some comments - please review when you get a chance. Thanks.

Thank you @chuehlien for some critical fixes 🎉

j-d-b commented on spec-content/enumerated-types/ArrowBoardPattern.md:
```diff
@@ -0,0 +1,32 @@
+`left-arrow-sequential` | Merge left represented by an arrow pattern (e.g. `<--`) that is displayed in a progressing sequence (e.g. `<` `<-` `<--` or `-` `--` `<--`).
+`left-chevron-static` | Merge left represented by a pattern of chevrons (e.g. `<<<`) that does not flash or move.
+`left-chevron-flashing` | Merge left represented by a pattern of chevrons (e.g. `<<<`) that that flashes on/off.
+`left-chevron-sequential` | Merge left represented by a pattern of chevrons that is displayed in a progressing sequence.
```
Great catch. It should be `chevron` (singular) to match Iowa's SABP and Ver-Mac's device feed, both which were inspiration for the enumerated type values.

sergebeaudry commented:
I would suggest to change the  **DynamicMessagesSign** object name to **ChangeableMessageSign** or simply **MessageSign**.

Those terms are in process to be refined by ASHTO, MUTCD and NEMA TS4.  currently they are mixed and they should not. 
The umbrella term is going to be CMS (ChangeableMessageSign) as currently used in MUTCD (chapter 2L)  

_Changeable Message Sign—a sign that is capable of displaying more than one message (one of
which might be a “blank” display), changeable manually, by remote control, or by automatic
control. Electronic-display changeable message signs are referred to as Dynamic Message
Signs in the National Intelligent Transportation Systems (ITS) Architecture and are referred
to as Variable Message Signs in the National Electrical Manufacturers Association (NEMA)
standards publication_.

Dynamic Message Sign term will be reserved in the next MUTCD revision for high definition CMS. 

The latest proposal at NEMA and ASHTO is based on this
![image](https://user-images.githubusercontent.com/54321625/139950731-700cb208-137e-46d3-b253-0af33ad84d55.png)

RHolt-RIDOT commented:
following up on @sergebeaudry's input above, I would support using the ChangeableMessageSign (CMS) object name.

j-d-b commented:
@sergebeaudry your comment is valuable but was added too late to change what was approved for v1.0 of the `SwzDeviceFeed` specification. If you think it is important enough to be worth changing the naming of the `DynamicMessageSign` object to `ChangeableMessageSign` for the next specification release, please [create an issue](https://github.com/usdot-jpo-ode/wzdx/issues/new/) for that discussion.

Thank you.

j-d-b commented:
I made changes to the file location and organization of the JSON schema to remove duplication between the schemas for the `WZDxFeed` and the `SwzDeviceFeed`. This will also facilitate the addition of the schema for the `RoadRestrictionFeed` in #205. Specifically:

- Moved the `schemas` directory to the project root for visibility and shorter URIs for schema `$id`s.
- Added a `4.0` subdirectory for the the WZDx version 4.0 schema. New releases should add a directory matching the version. This avoids having to have the version in every file name, again simplifying URIs. It also clarifies that the `SwzDeviceFeed` is part of WZDx v4.0, which is necessary since it shares objects (`FeedInfo` and `FeedDataSource`) with the `WZDxFeed`.
- Moved the schemas for WZDx objects that appear in both feeds (`FeedInfo`, `FeedDataSource`, `BoundingBox`) to separate files and referenced those files in the schema for each feed to avoid duplication.
- Updated the use of `oneOf` within the `RoadEventFeature` and `FieldDeviceFeature` `properties` to check the `core_details` `event_type` and `device_type` respectively to validate the rest of the object according to the type provided. It looks a bit messier, but it provides valuable functionality.

j-d-b commented:
I did a bit of refactoring of the project README to describe that WZDx now defines multiple feeds. There is more I want to do with regards to visibility of important information and clarifying what WZDx is, however, that is better implemented in a separate PR to avoid cluttering this one.

j-d-b commented:
I did two more things as part of helping clarify that there are now multiple "WZDx Feeds":

- Move the examples out of the `create-feed` directory and into the project root for visibility—after this `create-feed` contained only a README Markdown file.
- Move the `create-feed` README to `Creating_a_WZDx_Feed.md` for clarity and visibility.
- Remove duplicate content between the project root `README.md` and `Creating_a_WZDx_Feed.md`.
- Update references.

What we should do as part of the `v4.0-release` editorial changes is review the `Creating_a_WZDx_Feed.md` vs project `README.md` vs `/spec-content/README.md` and see if we can make each of those files have a clear message without much overlap to facilitate new users being able to understand what the repository/wzdx is defining and how to use it.

j-d-b commented on schemas/4.0/BoundingBox.json:
```diff
@@ -0,0 +1,11 @@
+{
+  "$id": "https://raw.githubusercontent.com/usdot-jpo-ode/wzdx/field-devices/schemas/4.0/BoundingBox.json",
```
Note all these `$id`s will have to be updated from `field-devices` to `main` for the v4.0-release.

j-d-b commented on spec-content/enumerated-types/LocationMethod.md:
```diff
@@ -13,7 +13,7 @@ Value | Description
 Property | Object
 --- | ---
-`location_method` | [WorkZoneRoadEvent](/spec-content/objects/WorkZoneRoadEvent.md)
+`location_method` | [FeedDataSource](/spec-content/objects/FeedDataSource.md)
```
`WorkZoneRoadEvent` is correct, not `FeedDataSource`

j-d-b commented on schemas/wzdx_v3.1_feed.json:
```diff
@@ -1,6 +1,6 @@
 
 {
-  "$id": "https://raw.githubusercontent.com/usdot-jpo-ode/wzdx/master/create-feed/schemas/wzdx_v3.1_feed.json",
+  "$id": "https://raw.githubusercontent.com/usdot-jpo-ode/wzdx/master/schemas/wzdx_v3.1_feed.json",
```
Fixed—they only need to apply once they get merged to `main` so I used that.

j-d-b commented on spec-content/README.md:
```diff
@@ -19,26 +19,44 @@ Each object is described by a table of properties, which are each defined by the
-This section provides a tabular list of all objects used in the WZDx specification.
+This section provides a tabular list of all objects used in the WZDx specification, sectioned by whether the object applies to the multiple feeds (shared), the `WZDxFeed`, or the `SwzDeviceFeed`.
 
+#### Shared
```
Done, but I think there is more organization that could be helpful.

mark-mockett reviewed (approved):
Great work on all of this @j-d-b. Other than the one typo in the FeedInfo schema, I didn't see anything that needed revision. I'm unfamiliar with some of the new schema techniques, but they seem to pass on the examples we've provided so far.

j-d-b commented:
@mark-mockett thanks for reviewing. We can discuss the schema in our meeting next week.
